Fixes: #24185 Cilium-operator crash on empty azure node pools

forgems opened the pull request:
Empty VirtualMachineScaleSets returns 404 on ListVirtualMachineScaleSetNetworkInterfaces if there are no nodes in it. Continue with next VirtualMachineScaleSet in that case.

Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Thanks for contributing!

Azure allows for empty user node pools. In that case ListVirtualMachineScaleSetNetworkInterfaces will return 404 when trying to get interfaces associated with given node pool. With this change it will continue with next VirtualMachineScaleSet if this is NotFound error.

Fixes: #24185
```release-note
Fix Cilium Operator from crashing when encountering empty node pools on Azure
```

tommyp1ckles commented on pkg/azure/api/api.go:
```diff
@@ -189,6 +190,10 @@ func (c *Client) vmssNetworkInterfaces(ctx context.Context) ([]network.Interface
 		c.metricsAPI.ObserveAPICall("Interfaces.ListVirtualMachineScaleSetNetworkInterfacesComplete", deriveStatus(err2), sinceStart.Seconds())
 		if err2 != nil {
+			// For scale set without any instances API will return not found. Then it can be skipped.
+			if v, ok := err2.(autorest.DetailedError); ok && v.StatusCode == http.StatusNotFound {
```
Makes sense, I did a bit more digging. Strangely, it seems like the Azure API only returns NotFound when the vmss was created as part of a AKS node-group, otherwise it returns a 200 with any empty list.

Could we document this idiosyncrasy in the comment for the next person looking at this.

forgems replied:
sure

forgems replied:
Done

forgems committed:
Fixes: #24185

Empty VirtualMachineScaleSets returns 404 on ListVirtualMachineScaleSetNetworkInterfaces if there are no nodes in it.
Continue with next VirtualMachineScaleSet in that case.

Signed-off-by: Marcin Swiderski <marcin@arangodb.com>
